src/coreclr/jit/emit.cpp

```diff
@@ -9276,40 +9276,62 @@ void emitter::emitRemoveLastInstruction()
     assert(emitLastIns != nullptr);
     assert(emitLastInsIG != nullptr);
 
-    JITDUMP("Removing saved instruction in %s:\n> ", emitLabelString(emitLastInsIG));
-    JITDUMPEXEC(dispIns(emitLastIns))
-
     // We should assert it's not a jmp, as that would require updating the jump lists, e.g. emitCurIGjmpList.
 
     BYTE*          lastInsActualStartAddr = (BYTE*)emitLastIns - m_debugInfoSize;
     unsigned short lastCodeSize           = (unsigned short)emitLastIns->idCodeSize();
 
-    // Check that a new buffer hasn't been create since the last instruction was emitted.
-    assert((emitCurIGfreeBase <= lastInsActualStartAddr) && (lastInsActualStartAddr < emitCurIGfreeEndp));
+    if ((emitCurIGfreeBase <= lastInsActualStartAddr) && (lastInsActualStartAddr < emitCurIGfreeEndp))
+    {
+        JITDUMP("Removing saved instruction in current IG %s:\n> ", emitLabelString(emitLastInsIG));
+        JITDUMPEXEC(emitDispIns(emitLastIns, /* isNew */ false, /* doffs */ false, /* asmfm */ false));
+
+        // The last instruction is in the current buffer. That means the current IG is non-empty.
+        assert(emitCurIGnonEmpty());
+        assert(lastInsActualStartAddr < emitCurIGfreeNext);
+        assert(emitCurIGinsCnt >= 1);
+        assert(emitCurIGsize >= emitLastIns->idCodeSize());
 
-    // Ensure the current IG is non-empty.
-    assert(emitCurIGnonEmpty());
-    assert(lastInsActualStartAddr < emitCurIGfreeNext);
-    assert(emitCurIGinsCnt >= 1);
-    assert(emitCurIGsize >= emitLastIns->idCodeSize());
+        size_t insSize = emitCurIGfreeNext - lastInsActualStartAddr;
```

Comment:
Minor nit: moving this insSize to directly before the memset (like in the else case), would make the then and else cases match better.